jax/_src/cudnn/fused_attention_stableHLO.py

```diff
@@ -0,0 +1,672 @@
+    if is_flash_attention:
+        activation = jnp.reshape(activation, (*batch_tuple, num_heads, q_seq_len))
+    else:
+        activation = jnp.reshape(activation, (*batch_tuple, num_heads, q_seq_len, kv_seq_len))
+    return (output, activation), out_bdims
+
+def _dot_product_attention_bwd_batcher(batched_args, batch_dims, *, scale, seed, dropout_rate, variadic_args, is_flash_attention, is_causal_mask):
+    _check_valid_batch_dims(batch_dims)
+    query, key, value, bias, mask, activation, fwd_output, grad_output = batched_args
+    query_bdim = batch_dims[0]
+    out_bdims = query_bdim, query_bdim, query_bdim
+
+    *batch_tuple, q_seq_len, num_heads, head_dim = query.shape
+    *_, kv_seq_len, _, _ = key.shape
+    batch = reduce(operator.mul, batch_tuple)
+    has_bias, has_mask = variadic_args
+    # reshape to 4D shape
+    query = jnp.reshape(query, (batch, q_seq_len, num_heads, head_dim))
+    key = jnp.reshape(key, (batch, kv_seq_len, num_heads, head_dim))
+    value = jnp.reshape(value, (batch, kv_seq_len, num_heads, head_dim))
+    if has_bias:
+        bias = jnp.reshape(bias, (batch, num_heads, q_seq_len, kv_seq_len))
+    if has_mask:
+        mask = jnp.reshape(mask, (batch, num_heads, q_seq_len, kv_seq_len))
+    if is_flash_attention:
+        activation = jnp.reshape(activation, (batch, num_heads, q_seq_len))
+    else:
+        activation = jnp.reshape(activation, (batch, num_heads, q_seq_len, kv_seq_len))
+    fwd_output = jnp.reshape(fwd_output, (batch, q_seq_len, num_heads, head_dim))
+    grad_output = jnp.reshape(grad_output, (batch, q_seq_len, num_heads, head_dim))
+
+    grad_query, grad_key, grad_value = _dot_product_attention_bwd_p_wrapper.bind(
+        query, key, value, bias,
+        mask, activation, fwd_output, grad_output,
+        scale=scale, seed=seed, dropout_rate=dropout_rate,
+        variadic_args=variadic_args, is_flash_attention=is_flash_attention,
+        is_causal_mask=is_causal_mask)
+    
+    # reshape to original shape
+    grad_query = jnp.reshape(grad_query, (*batch_tuple, q_seq_len, num_heads, head_dim))
+    grad_key = jnp.reshape(grad_key, (*batch_tuple, kv_seq_len, num_heads, head_dim))
+    grad_value = jnp.reshape(grad_value, (*batch_tuple, kv_seq_len, num_heads, head_dim))
+    grads = (grad_query, grad_key, grad_value)
+    return grads, out_bdims
+
+# custom partitioning
+def _get_padded_spec(arg_info):
+    spec = None if arg_info.sharding is None else arg_info.sharding.spec
+    ndim = arg_info.ndim
+    if spec is None:
+        return (None,) * ndim
+    assert len(spec) <= ndim
+    return spec + (None,) * (ndim - len(spec))
+    
+# fwd custom partition
+_dot_product_attention_fwd_lower = custom_partitioning(_dot_product_attention_fwd_impl, static_argnums=(5,6,7,8,9,10))
+def _dot_product_attention_fwd_infer_sharding_from_operands(scale, seed, dropout_rate, variadic_args, is_flash_attention, is_causal_mask, mesh, arg_shapes, result_shape):
```

Comment:
This function shares a lot of code with the `partition` function. Can the logic that computes shardings be shared via a helper function?